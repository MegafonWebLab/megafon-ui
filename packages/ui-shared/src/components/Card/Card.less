@import '~@megafon/ui-core/styles/base.less';
@b: e('.mfui-beta-card');

@horizontalIndent: 20px;
@horizontalIndentSM: 16px;
@verticalIndent: 32px;
@verticalIndentSM: 24px;

@{b} {
    display: flex;
    border-radius: 8px;
    box-shadow: 0 0.25px 3px rgba(0, 0, 0, 0.25);
    background-color: @clearWhite;
    overflow: hidden;

    &_href {
        * {
            text-decoration: none;
        }

        &:hover {
            box-shadow: 0 8px 28px rgba(0, 0, 0, 0.1);
        }
    }

    &__inner {
        display: flex;
        flex-direction: column;
        flex-grow: 0;
        flex-shrink: 1;
        width: 100%;
        box-sizing: border-box;
        padding: @verticalIndent @horizontalIndent;

        @media @mobileSM {
            padding: @verticalIndentSM @horizontalIndentSM;
        }
    }

    &__title {
        margin-bottom: 12px;
    }

    &__btns-wrapper {
        display: flex;
        flex-wrap: wrap;
        justify-content: space-evenly;
        align-items: center;
        padding-top: 12px;
        margin-top: auto;
    }

    &__btns-wrapper_left-align {
        justify-content: flex-start;
    }

    &__fake-link {
        color: @link;
    }

    &__button,
    &__link,
    &__fake-link {
        margin: 12px 10px;
    }

    &__pic-wrapper {
        margin-bottom: 32px;

        @media @mobileSM {
            margin-bottom: 24px;
        }
    }

    &__pic-wrapper &__img {
        max-width: 100%;
    }

    &__pic-wrapper svg {
        width: 40px;
        height: 40px;
    }

    &__pic-wrapper_img {
        display: flex;
        justify-content: center;
        align-items: center;
        margin-top: -(@verticalIndent);
        margin-left: -(@horizontalIndent);
        margin-right: -(@horizontalIndent);

        @media @mobileSM {
            margin-top: -(@verticalIndentSM);
            margin-left: -(@horizontalIndentSM);
            margin-right: -(@horizontalIndentSM);
        }
    }

    &__pic-wrapper_object-fit_fill img {
        width: 100%;
        height: auto;
    }

    &__pic-wrapper_object-fit_contain img {
        width: auto;
        height: 100%;
    }

    &_full-height {
        height: 100%;
    }
}
