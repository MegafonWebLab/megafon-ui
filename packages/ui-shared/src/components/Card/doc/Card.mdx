---
name: Card
menu: Конструктор
route: components/card
---

import Example from './Card.example.mdx';
import Props from './Card.props.mdx';

# Card

<CopyToClipboardBox text="import { Card } from '@megafon/ui-shared';" />

<DoczTabs>
    <Example />
    <Props />
</DoczTabs>
