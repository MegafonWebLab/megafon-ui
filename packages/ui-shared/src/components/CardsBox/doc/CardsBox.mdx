---
name: CardsBox
menu: Конструктор
route: components/cards-box
---

import { Props } from 'docz';
import CardsBox from '../CardsBox';
import Example from './CardsBox.example.mdx'

<Example />
