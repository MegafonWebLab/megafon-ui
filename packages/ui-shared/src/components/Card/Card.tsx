import React, { Ref } from 'react';
import './Card.less';
import { cnCreate, Header, Paragraph, Button, TextLink, Link, dataAttrs as filterDataAttrs } from '@megafon/ui-core';
import PropTypes from 'prop-types';

interface IButton {
    title: string;
    href: string;
}

interface ILink {
    title: string;
    href?: string;
}

export const ObjectFit = {
    FILL: 'fill',
    CONTAIN: 'contain',
} as const;

type ObjectFitType = typeof ObjectFit[keyof typeof ObjectFit];

export interface ICard {
    /** Дата атрибуты для корневого элемента */
    dataAttrs?: { [key: string]: string };
    /** Дополнительный класс корневого элемента */
    className?: string;
    /** Дополнительные классы для корневого и внутренних элементов */
    classes?: {
        root?: string;
        button?: string;
        link?: string;
        inner?: string;
    };
    /** Ссылка на корневой элемент */
    rootRef?: Ref<HTMLDivElement>;
    /** Изображение в карточке */
    imageSrc?: string;
    /** Иконка в карточке */
    svgSrc?: React.ReactNode;
    /** Заголовок карточки */
    title: string;
    /** Текст карточки */
    text: string;
    /** Данные для кнопки */
    button?: IButton;
    /** Данные для ссылки */
    link?: ILink;
    /** Расположение кнопки/ссылки по левой стороне */
    isLeftHAlign?: boolean;
    /** Высота корневого элемента 100% */
    isFullHeight?: boolean;
    /** Ссылка для всей карточки */
    href?: string;
    /** Режим позиционирования изображения */
    objectFit?: ObjectFitType;
}

const cn = cnCreate('mfui-beta-card');
const Card: React.FC<ICard> = ({
    dataAttrs,
    className,
    classes = {},
    rootRef,
    imageSrc,
    svgSrc,
    title,
    text,
    button,
    link,
    isLeftHAlign = false,
    isFullHeight = false,
    href,
    objectFit = 'fill',
}) => {
    const isAlignAvailable = !button || !link;
    const isCardLink = !!href;
    const isRenderBtn = !!button && !isCardLink;
    const Element = href ? Link : 'div';

    const renderImage = React.useCallback(() => {
        switch (true) {
            case !!imageSrc: {
                return (
                    <div className={cn('pic-wrapper', { 'object-fit': objectFit, 'img': true })}>
                        <img className={cn('img')} src={imageSrc} />
                    </div>
                );
            }

            case !!svgSrc: {
                return (
                    <div className={cn('pic-wrapper')}>
                        {svgSrc}
                    </div>
                );
            }

            default:
                return null;
        }
    }, [imageSrc, svgSrc, objectFit]);

    const renderLink = React.useCallback(({ href: linkHref, title: linkTitle }) => {
        const isFakeLink = !linkHref;

        if (isFakeLink || isCardLink) {
            return <span className={cn('fake-link')}>{linkTitle}</span>;
        }

        return (
            <TextLink className={cn('link', [classes.link])} href={linkHref}>{linkTitle}</TextLink>
        );
    }, [isCardLink]);

    const renderBtn = React.useCallback(({ href: btnHref, title: btnTitle}) => (
        <Button className={cn('button', [classes.button])} href={btnHref}>{btnTitle}</Button>
    ), []);

    return (
        <div
            {...filterDataAttrs(dataAttrs)}
            className={cn('', { 'href': !!href, 'full-height': isFullHeight }, [className, classes.root])}
            ref={rootRef}
        >
            <Element href={href} className={cn('inner', [classes.inner])}>
                <>
                    {renderImage()}
                    <Header as="h3" className={cn('title')}>{title}</Header>
                    <Paragraph hasMargin={false}>{text}</Paragraph>
                    <div className={cn('btns-wrapper', { 'left-align': isAlignAvailable && isLeftHAlign })}>
                        {isRenderBtn && renderBtn(button)}
                        {link && renderLink(link)}
                    </div>
                </>
            </Element>
        </div>
    );
};

Card.propTypes = {
    dataAttrs: PropTypes.objectOf(PropTypes.string.isRequired),
    className: PropTypes.string,
    classes: PropTypes.shape({
        root: PropTypes.string,
        button: PropTypes.string,
        link: PropTypes.string,
        inner: PropTypes.string,
    }),
    rootRef: PropTypes.oneOfType([
        PropTypes.func,
        PropTypes.oneOfType([PropTypes.shape({ current: PropTypes.elementType }), PropTypes.any ]),
    ]),
    imageSrc: PropTypes.string,
    svgSrc: PropTypes.node,
    title: PropTypes.string.isRequired,
    text: PropTypes.string.isRequired,
    button: PropTypes.shape({
        title: PropTypes.string.isRequired,
        href: PropTypes.string.isRequired,
    }),
    link: PropTypes.shape({
        title: PropTypes.string.isRequired,
        href: PropTypes.string,
    }),
    isLeftHAlign: PropTypes.bool,
    isFullHeight: PropTypes.bool,
    href: PropTypes.string,
    objectFit: PropTypes.oneOf(Object.values(ObjectFit)),
};

export default Card;
