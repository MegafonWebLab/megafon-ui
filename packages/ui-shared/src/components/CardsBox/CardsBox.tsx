import * as React from 'react';
import * as PropTypes from 'prop-types';
import throttle from 'lodash.throttle';
import { cnCreate, Grid, GridColumn, Carousel, breakpoints, throttleTime } from '@megafon/ui-core';
import { ICard } from '../Card/Card';

type SlidesSettingsType = Pick<React.ComponentProps<typeof Carousel>, 'slidesSettings'>['slidesSettings'];

const MAX_CARDS_COUNT_IN_GRID_ON_MOBILE = 2;

const SlidesSettings: SlidesSettingsType = {
    [breakpoints.mobileSmallStart]: {
        slidesPerView: 1,
        spaceBetween: 16,
    },
    [breakpoints.mobileBigStart]: {
        slidesPerView: 2,
        spaceBetween: 20,
    },
};

interface ICardsBoxProps {
    children: Array<React.ReactElement<ICard>> | React.ReactElement<ICard>;
}

const cn = cnCreate('mfui-beta-cards-box');
const CardsBox: React.FC<ICardsBoxProps> = ({ children }) => {
    const [isMobile, setIsMobile] = React.useState(false);
    const itemsCount = React.Children.count(children);
    const isRenderCarousel = isMobile && itemsCount > MAX_CARDS_COUNT_IN_GRID_ON_MOBILE;

    const renderGrid = React.useCallback(() => (
        <Grid guttersBottom="medium" guttersLeft="medium">
            {
                React.Children.map(children, child => (
                    <GridColumn all="4" mobile="12">
                        {child}
                    </GridColumn>
                ))
            }
        </Grid>
    ), [children]);

    const renderCarousel = React.useCallback(() => (
        <Carousel slidesSettings={SlidesSettings}>{children}</Carousel>
    ), [children]);

    React.useEffect(() => {
        const resizeHandler = () =>
            window.innerWidth <= breakpoints.mobileBigEnd ? setIsMobile(true) : setIsMobile(false);
        const resizeHandlerThrottled = throttle(resizeHandler, throttleTime.resize);

        resizeHandler();
        window.addEventListener('resize', resizeHandlerThrottled);

        return () => {
            window.removeEventListener('resize', resizeHandlerThrottled);
        };
    }, []);

    return (
        <div className={cn()}>
            {isRenderCarousel ? renderCarousel() : renderGrid()}
        </div>
    );
};

CardsBox.propTypes = {
    children: PropTypes.oneOfType([
        PropTypes.arrayOf(PropTypes.element.isRequired),
        PropTypes.element,
    ]).isRequired,
};

export default CardsBox;
