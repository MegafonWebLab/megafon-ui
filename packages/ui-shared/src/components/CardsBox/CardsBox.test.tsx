import * as React from 'react';
import { shallow, mount } from 'enzyme';
import CardsBox from './CardsBox';
import Card, { ICard } from '../Card/Card';

const cardProps: ICard = {
    title: 'title',
    text: 'text',
    button: {
        title: 'button-title',
        href: 'button-href',
    },
    link: {
        title: 'href-title',
        href: 'href',
    },
};

type LocalWindowType = Omit<Window, 'innerWidth'> & {
    innerWidth: number;
};

jest.mock('@megafon/ui-core/dist/lib/components/Carousel/Carousel', () => {
    const CarouselMock = (props): React.ReactNode => props.children;

    return CarouselMock;
});

describe('CardsBox', () => {
    it('render component', () => {
        const wrapper = shallow(<CardsBox><Card {...cardProps} /></CardsBox>);
        expect(wrapper).toMatchSnapshot();
    });

    it('render on mobile resolution with carousel', () => {
        const localWindow = window as LocalWindowType;
        const windowInnerWidth = window.innerWidth;

        localWindow.innerWidth = 320;

        const wrapper = mount(
            <CardsBox>
                <Card {...cardProps} />
                <Card {...cardProps} />
                <Card {...cardProps} />
            </CardsBox>
        );

        expect(wrapper).toMatchSnapshot();

        localWindow.innerWidth = windowInnerWidth;
    });
});
