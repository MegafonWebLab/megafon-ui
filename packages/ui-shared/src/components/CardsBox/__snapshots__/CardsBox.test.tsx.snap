// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardsBox render component 1`] = `
<div
  className="mfui-beta-cards-box"
>
  <Grid
    guttersBottom="medium"
    guttersLeft="medium"
    multiRow={true}
  >
    <GridColumn
      all="4"
      flex={false}
      grow={false}
      key=".0"
      mobile="12"
      orderAll="0"
      orderDesktop="0"
      orderMobile="0"
      orderTablet="0"
      orderWide="0"
    >
      <Card
        button={
          Object {
            "href": "button-href",
            "title": "button-title",
          }
        }
        link={
          Object {
            "href": "href",
            "title": "href-title",
          }
        }
        text="text"
        title="title"
      />
    </GridColumn>
  </Grid>
</div>
`;

exports[`CardsBox render on mobile resolution with carousel 1`] = `
<CardsBox>
  <div
    className="mfui-beta-cards-box"
  >
    <CarouselMock
      slidesSettings={
        Object {
          "0": Object {
            "slidesPerView": 1,
            "spaceBetween": 16,
          },
          "768": Object {
            "slidesPerView": 2,
            "spaceBetween": 20,
          },
        }
      }
    >
      <Card
        button={
          Object {
            "href": "button-href",
            "title": "button-title",
          }
        }
        link={
          Object {
            "href": "href",
            "title": "href-title",
          }
        }
        text="text"
        title="title"
      >
        <div
          className="mfui-beta-card"
        >
          <div
            className="mfui-beta-card__inner"
          >
            <Header
              as="h3"
              className="mfui-beta-card__title"
              color="black"
              hAlign="inherit"
            >
              <h3
                className="mfui-beta-header mfui-beta-header_color_black mfui-beta-header_level_h3 mfui-beta-header_h-align_inherit mfui-beta-card__title"
              >
                title
              </h3>
            </Header>
            <Paragraph
              hasMargin={false}
            >
              <p
                className="mfui-beta-paragraph mfui-beta-paragraph_size_regular mfui-beta-paragraph_color_freshAsphalt"
              >
                text
              </p>
            </Paragraph>
            <div
              className="mfui-beta-card__btns-wrapper"
            >
              <Button
                className="mfui-beta-card__button"
                href="button-href"
              >
                <a
                  className="mfui-beta-button mfui-beta-button_type_primary mfui-beta-button_theme_green mfui-beta-button_size-all_medium mfui-beta-button_no-touch mfui-beta-card__button"
                  disabled={false}
                  href="button-href"
                  onClick={[Function]}
                >
                  <div
                    className="mfui-beta-button__inner"
                  >
                    <div
                      className="mfui-beta-button__content"
                    >
                      <span
                        className="mfui-beta-button__text"
                      >
                        button-title
                      </span>
                    </div>
                  </div>
                </a>
              </Button>
              <TextLink
                className="mfui-beta-card__link"
                color="blue"
                href="href"
                underlineStyle="solid"
                underlineVisibility="hover"
              >
                <Link
                  className="mfui-beta-text-link mfui-beta-text-link_underline-visibility_hover mfui-beta-text-link_underline-style_solid mfui-beta-text-link_color_blue mfui-beta-card__link"
                  href="href"
                >
                  <a
                    className="mfui-beta-text-link mfui-beta-text-link_underline-visibility_hover mfui-beta-text-link_underline-style_solid mfui-beta-text-link_color_blue mfui-beta-card__link"
                    href="href"
                  >
                    href-title
                  </a>
                </Link>
              </TextLink>
            </div>
          </div>
        </div>
      </Card>
      <Card
        button={
          Object {
            "href": "button-href",
            "title": "button-title",
          }
        }
        link={
          Object {
            "href": "href",
            "title": "href-title",
          }
        }
        text="text"
        title="title"
      >
        <div
          className="mfui-beta-card"
        >
          <div
            className="mfui-beta-card__inner"
          >
            <Header
              as="h3"
              className="mfui-beta-card__title"
              color="black"
              hAlign="inherit"
            >
              <h3
                className="mfui-beta-header mfui-beta-header_color_black mfui-beta-header_level_h3 mfui-beta-header_h-align_inherit mfui-beta-card__title"
              >
                title
              </h3>
            </Header>
            <Paragraph
              hasMargin={false}
            >
              <p
                className="mfui-beta-paragraph mfui-beta-paragraph_size_regular mfui-beta-paragraph_color_freshAsphalt"
              >
                text
              </p>
            </Paragraph>
            <div
              className="mfui-beta-card__btns-wrapper"
            >
              <Button
                className="mfui-beta-card__button"
                href="button-href"
              >
                <a
                  className="mfui-beta-button mfui-beta-button_type_primary mfui-beta-button_theme_green mfui-beta-button_size-all_medium mfui-beta-button_no-touch mfui-beta-card__button"
                  disabled={false}
                  href="button-href"
                  onClick={[Function]}
                >
                  <div
                    className="mfui-beta-button__inner"
                  >
                    <div
                      className="mfui-beta-button__content"
                    >
                      <span
                        className="mfui-beta-button__text"
                      >
                        button-title
                      </span>
                    </div>
                  </div>
                </a>
              </Button>
              <TextLink
                className="mfui-beta-card__link"
                color="blue"
                href="href"
                underlineStyle="solid"
                underlineVisibility="hover"
              >
                <Link
                  className="mfui-beta-text-link mfui-beta-text-link_underline-visibility_hover mfui-beta-text-link_underline-style_solid mfui-beta-text-link_color_blue mfui-beta-card__link"
                  href="href"
                >
                  <a
                    className="mfui-beta-text-link mfui-beta-text-link_underline-visibility_hover mfui-beta-text-link_underline-style_solid mfui-beta-text-link_color_blue mfui-beta-card__link"
                    href="href"
                  >
                    href-title
                  </a>
                </Link>
              </TextLink>
            </div>
          </div>
        </div>
      </Card>
      <Card
        button={
          Object {
            "href": "button-href",
            "title": "button-title",
          }
        }
        link={
          Object {
            "href": "href",
            "title": "href-title",
          }
        }
        text="text"
        title="title"
      >
        <div
          className="mfui-beta-card"
        >
          <div
            className="mfui-beta-card__inner"
          >
            <Header
              as="h3"
              className="mfui-beta-card__title"
              color="black"
              hAlign="inherit"
            >
              <h3
                className="mfui-beta-header mfui-beta-header_color_black mfui-beta-header_level_h3 mfui-beta-header_h-align_inherit mfui-beta-card__title"
              >
                title
              </h3>
            </Header>
            <Paragraph
              hasMargin={false}
            >
              <p
                className="mfui-beta-paragraph mfui-beta-paragraph_size_regular mfui-beta-paragraph_color_freshAsphalt"
              >
                text
              </p>
            </Paragraph>
            <div
              className="mfui-beta-card__btns-wrapper"
            >
              <Button
                className="mfui-beta-card__button"
                href="button-href"
              >
                <a
                  className="mfui-beta-button mfui-beta-button_type_primary mfui-beta-button_theme_green mfui-beta-button_size-all_medium mfui-beta-button_no-touch mfui-beta-card__button"
                  disabled={false}
                  href="button-href"
                  onClick={[Function]}
                >
                  <div
                    className="mfui-beta-button__inner"
                  >
                    <div
                      className="mfui-beta-button__content"
                    >
                      <span
                        className="mfui-beta-button__text"
                      >
                        button-title
                      </span>
                    </div>
                  </div>
                </a>
              </Button>
              <TextLink
                className="mfui-beta-card__link"
                color="blue"
                href="href"
                underlineStyle="solid"
                underlineVisibility="hover"
              >
                <Link
                  className="mfui-beta-text-link mfui-beta-text-link_underline-visibility_hover mfui-beta-text-link_underline-style_solid mfui-beta-text-link_color_blue mfui-beta-card__link"
                  href="href"
                >
                  <a
                    className="mfui-beta-text-link mfui-beta-text-link_underline-visibility_hover mfui-beta-text-link_underline-style_solid mfui-beta-text-link_color_blue mfui-beta-card__link"
                    href="href"
                  >
                    href-title
                  </a>
                </Link>
              </TextLink>
            </div>
          </div>
        </div>
      </Card>
    </CarouselMock>
  </div>
</CardsBox>
`;
